### PR TITLE
Allow users to define vm naming logic

### PR DESCRIPTION
**Pull request by erkanerol**
## Description

Now, VMs are named as `machine.Name`. In some cases, we need to name VMs with a custom logic.

This PR adds a new field into VCDMachine CR so that users can define templates to name VMs by using [Go templates](https://pkg.go.dev/text/template) and [Sprig functions](https://github.com/Masterminds/sprig). `.machine` and `.vcdMachine` will refer to Machine and VCDMachine CRs in the templates.

***Example***
```
spec:
  vmNamingTemplate: 'mycustomprefix{{.machine.Name |sha256sum | trunc 7}}'
```

***Important Points***
This field must be immutable. Otherwise, users can change it and the controller will not be able to know existing VMs. 
Immutability is enforced by the new validating webhook for VCDMachine. 

## Checklist
- [x] tested locally
- [ ] updated any relevant dependencies
- [ ] updated any relevant documentation or examples

## API Changes
Are there API changes?
- [x] Yes
- [ ] No

If yes, please fill in the below

1. Updated conversions?
    - [ ] Yes
    - [ ] No
    - [ ] N/A
2. Updated CRDs?
    - [ ] Yes
    - [ ] No
    - [ ] N/A
3. Updated infrastructure-components.yaml?
    - [ ] Yes
    - [ ] No
    - [ ] N/A
4. Updated `./examples/capi-quickstart.yaml`?
    - [ ] Yes
    - [ ] No
    - [ ] N/A
5. Updated necessary files under `./infrastructure-vcd/v1.0.0/`?
   - [ ] Yes
   - [ ] No
   - [ ] N/A

- - -
This change is [Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/369)

### CODE REVIEW

**Comment by erkanerol** [comment]
@arunmk I added an immutability check to the webhook. This is ready for review.

File: api/v1beta1/vcdmachine_types.go [hunk 1]
```diff
@@ -77,6 +77,12 @@ type VCDMachineSpec struct {
 	// VCDClusterSpec.OvdcNetwork is always attached regardless of this field.
 	// +optional
 	ExtraOvdcNetworks []string `json:"extraOvdcNetworks"`
+
+	// VmNamingTemplate is go template to generate VM names based on Machine and VCDMachine CRs.
+	// Functions of Sprig library are supported. See https://github.com/Masterminds/sprig
+	// Immutable field. machine.Name is used as VM name when this field is empty.
+	// +optional
+	VmNamingTemplate string `json:"vmNamingTemplate"`
```
Comment (arunmk) [inline api/v1beta1/vcdmachine_types.go:85]: This should also be immutable. Can you please add a comment mentioning it

**Review by arunmk — APPROVED** [review]
lgtm. I have only one ask of a comment

File: controllers/vcdmachine_controller.go [hunk 2]
```diff
@@ -441,7 +449,10 @@ func (r *VCDMachineReconciler) reconcileNormal(ctx context.Context, cluster *clu
 		cloudInitInput.HTTPProxy = vcdCluster.Spec.ProxyConfigSpec.HTTPProxy
 		cloudInitInput.HTTPSProxy = vcdCluster.Spec.ProxyConfigSpec.HTTPSProxy
 		cloudInitInput.NoProxy = vcdCluster.Spec.ProxyConfigSpec.NoProxy
-		cloudInitInput.MachineName = machine.Name
+		cloudInitInput.MachineName = vmName
+		if err != nil {
```
Comment (Anirudh9794) [inline controllers/vcdmachine_controller.go:453]: It is not sure what error is being checked here. Can you please make sure this is not added by mistake?

**Review by Anirudh9794 — APPROVED** [review]
Apart from the error check, which seems like which was added by mistake, the rest of the PR looks good to me.

File: controllers/vcdmachine_controller.go (hunk 2 above)
Comment (arunmk) [reply controllers/vcdmachine_controller.go:453]: @erkanerol could you please handle this? we should be able to merge immediately after it

Comment (erkanerol) [reply controllers/vcdmachine_controller.go]: It is leftover after a refactoring. Just deleted the check. Thanks for catching this.

**Comment by erkanerol** [comment]
@arunmk @Anirudh9794 I rebased the PR. FYI.

File: api/v1beta1/vcdmachine_types.go [hunk 3]
```diff
@@ -76,13 +76,13 @@ type VCDMachineSpec struct {
...
 	// +optional
-	ExtraOvdcNetworks []string `json:"extraOvdcNetworks"`
+	ExtraOvdcNetworks []string `json:"extraOvdcNetworks,omitempty"`
 
 	// VmNamingTemplate is go template to generate VM names based on Machine and VCDMachine CRs.
 	// Functions of Sprig library are supported. See https://github.com/Masterminds/sprig
 	// Immutable field. machine.Name is used as VM name when this field is empty.
 	// +optional
-	VmNamingTemplate string `json:"vmNamingTemplate"`
+	VmNamingTemplate string `json:"vmNamingTemplate,omitempty"`
```
Comment (erkanerol) [inline api/v1beta1/vcdmachine_types.go:85]: Users should be able to create VCDMachine CRs without providing empty values for these fields. Missing omitempty tags lead to errors in CAPI controllers while creating VCDMachine by using VCDMachineTemplate.

**Comment by arunmk** [comment]
I'll merge this since it has 2 approvals.

**Pull request merged**